### PR TITLE
Add net8.0 TFM to source-built packages

### PR DESCRIPTION
**Pull request by jjonescz**
Resolves https://github.com/dotnet/roslyn/issues/66930.

### CODE REVIEW

File: src/Compilers/VisualBasic/Portable/Microsoft.CodeAnalysis.VisualBasic.vbproj [hunk 1]
```diff
@@ -4,6 +4,7 @@
   <PropertyGroup>
     <OutputType>Library</OutputType>
     <TargetFrameworks>net6.0;netstandard2.0</TargetFrameworks>
+    <TargetFrameworks Condition="'$(DotNetBuildFromSource)' == 'true'">net8.0;$(TargetFrameworks)</TargetFrameworks>
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/Microsoft.CodeAnalysis.VisualBasic.vbproj:7]: Rather than put this condition in all these files I wonder if we should centralize a bit. Maybe in `Settings.props` have 

```xml
    <SourceBuildTargetFrameworks>net6.0</SourceBuildTargetFrameworks>
    <TargetFrameworks Condition="'$(DotNetBuildFromSource)' == 'true'">net8.0;net6.0</SourceBuildTargetFrameworks>
```

Then we can just use the `$(SourceBuildtargetFrameworks)` in the project files. The other advantage is we could further condition that we disable the `net8.0` when outside CI / official build. That way it doesn't build in the developer inner loop. 

I really don't like the name of the property though, open to pretty much any other suggestion 😄 

@333fred, @chsienki

Comment (333fred) [reply src/Compilers/VisualBasic/Portable/Microsoft.CodeAnalysis.VisualBasic.vbproj:7]: Well, see, I'm not sure whether I should be suggesting an alternative to `SourceBuildTargetFrameworks`, `TargetFrameworks`, or `SourceBuildtargetFrameworks` 😛

Comment (jjonescz) [reply src/Compilers/VisualBasic/Portable/Microsoft.CodeAnalysis.VisualBasic.vbproj:7]: Done. Btw, some projects packaged as part of source build target only `netstandard2.0`, they didn't have `net6.0`, so I didn't add `net8.0` to them, hope that's correct.

File: src/Compilers/CSharp/csc/AnyCpu/csc.csproj [hunk 2]
```diff
@@ -3,8 +3,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFrameworks>net6.0;net472</TargetFrameworks>
-    <TargetFrameworks Condition="'$(DotNetBuildFromSource)' == 'true'">net8.0;$(TargetFrameworks)</TargetFrameworks>
+    <TargetFrameworks>$(SourceBuildTargetFrameworks);net472</TargetFrameworks>
```
Comment (333fred) [inline src/Compilers/CSharp/csc/AnyCpu/csc.csproj:6]: I wonder whether we should be building net472 for source-build at all?

Comment (jjonescz) [reply src/Compilers/CSharp/csc/AnyCpu/csc.csproj:6]: You mean to speed the build up? Then we don't have to build net6.0 either, only net8.0?

Comment (333fred) [reply src/Compilers/CSharp/csc/AnyCpu/csc.csproj:6]: Yeah. Especially the net6.0 targets, I can only assume that this will cause pain in source-build. If they could easily build net6.0, we wouldn't need to update in the first place.

File: eng/targets/Imports.BeforeArcade.targets [hunk 3]
```diff
@@ -17,9 +17,11 @@
   <ItemGroup Condition="'$(DotNetBuildFromSource)' == 'true'">
     <KnownFrameworkReference Update="Microsoft.NETCore.App">
       <TargetingPackVersion Condition="'%(TargetFramework)' == 'net6.0'">6.0.0</TargetingPackVersion>
+      <TargetingPackVersion Condition="'%(TargetFramework)' == 'net8.0'">8.0.0-preview.3.23165.3</TargetingPackVersion>
     </KnownFrameworkReference>
     <KnownFrameworkReference Update="Microsoft.AspNetCore.App">
       <TargetingPackVersion Condition="'%(TargetFramework)' == 'net6.0'">6.0.0</TargetingPackVersion>
+      <TargetingPackVersion Condition="'%(TargetFramework)' == 'net8.0'">8.0.0-preview.3.23164.14</TargetingPackVersion>
```
Comment (jaredpar) [inline eng/targets/Imports.BeforeArcade.targets:24]: Are these values that we manually maintain or is this being updated by tooling?

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs [hunk 4]
```diff
@@ -9,6 +9,7 @@
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
 using System.IO;
+using System.Linq;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs:12]: Why was this change needed?

File: src/NuGet/Microsoft.Net.Compilers.Toolset/CoreClrCompilerArtifacts.targets [hunk 5]
```diff
@@ -3,37 +3,37 @@
...
+      <CoreClrCompilerBinArtifact Include="$(ArtifactsBinDir)VBCSCompiler\$(Configuration)\$(TargetFramework)\publish\VBCSCompiler.deps.json" />
+      <CoreClrCompilerBinArtifact Include="$(ArtifactsBinDir)VBCSCompiler\$(Configuration)\$(TargetFramework)\publish\VBCSCompiler.runtimeconfig.json" />
 
       <!-- References that are either not in the target framework or are a higher version -->
       <!-- N.B.: The backslashes below cannot be replaced with forward slashes.
          https://github.com/NuGet/Home/issues/3584 -->
-      <CoreClrCompilerBinArtifact Include="$(ArtifactsBinDir)csc\$(Configuration)\net6.0\publish\System.*.dll"/>
-      <CoreClrCompilerBinRuntimesArtifact Include="$(ArtifactsBinDir)csc\$(Configuration)\net6.0\publish\runtimes\**" />
+      <CoreClrCompilerBinArtifact Include="$(ArtifactsBinDir)csc\$(Configuration)\$(TargetFramework)\publish\System.*.dll"/>
+      <CoreClrCompilerBinRuntimesArtifact Include="$(ArtifactsBinDir)csc\$(Configuration)\$(TargetFramework)\publish\runtimes\**" />
```
Comment (jaredpar) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/CoreClrCompilerArtifacts.targets:36]: Well that is a clever way of doing it that I'm kicking myself for not thinking of originally :smile:

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CSharpSyntaxGenerator.csproj [hunk 6]
```diff
@@ -5,7 +5,8 @@
     <RootNamespace>CSharpSyntaxGenerator</RootNamespace>
     <!-- We build this against net7.0 to build the console app version, and against netstandard2.0 for
          the Source Generator version. -->
-    <TargetFrameworks>net7.0;netstandard2.0</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0</TargetFrameworks>
+    <TargetFrameworks Condition="'$(DotNetBuildFromSource)' != 'true'">net7.0;$(TargetFrameworks)</TargetFrameworks>
```
Comment (jaredpar) [inline src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CSharpSyntaxGenerator.csproj:9]: Is there a reason why we _need_ `net7.0` here? I think we could simplify this by just changing to 

```xml
<TargetFrameworks>net8.0;netstandard2.0</TargetFrameworks>
```

Comment (333fred) [reply src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CSharpSyntaxGenerator.csproj:9]: I don't think so, this could very easily be 8.0.

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs (hunk 4 above)
Comment (jjonescz) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs:12]: Because net8.0 has `IEnumerable<T>.Order` extension method, so I disabled the one from Roslyn utilities in net8.0, but here the `EnumerableExtensions` weren't visible.

<samp>roslyn\src\Compilers\Core\Portable\CommandLine\AnalyzerConfigSet.cs(553,58): error CS1061: 'IEnumerable<string>' does not contain a definition for 'Order' and no accessible extension method 'Order' accepting a first argument of type 'IEnumerable<string>' could be found (are you missing a using directive or an assembly reference?) [roslyn\src\Compilers\Core\Portable\Microsoft.CodeAnalysis.csproj::TargetFramework=net8.0]</samp>

File: src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CSharpSyntaxGenerator.csproj (hunk 6 above)
Comment (jjonescz) [reply src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/CSharpSyntaxGenerator.csproj:9]: This can be simplified as you suggest, but needed to make `generate-compiler-code.ps1` more complex...

File: eng/targets/Settings.props [hunk 7]
```diff
@@ -53,6 +53,18 @@
     <EnforceExtendedAnalyzerRules Condition="'$(IsAnalyzer)' == 'true'">true</EnforceExtendedAnalyzerRules>
   </PropertyGroup>
 
+  <PropertyGroup>
```
[the remark this answers is not in the document]
Comment (jjonescz) [reply eng/targets/Settings.props:56]: Alright, I changed the logic to be like this:

| Environment | `SourceBuildTargetFrameworks` | `SourceBuildTargetFrameworksNetFx` |
|--------|--------|--------|
| CI | net8.0; net6.0 | net8.0; net6.0; net472 |
| Local | net6.0 | net6.0; net472 |
| Source Build | net8.0 | net8.0 |

File: src/Compilers/Core/Portable/Microsoft.CodeAnalysis.csproj [hunk 8]
```diff
@@ -29,8 +29,10 @@
 
       Note: PrivateAssets="ContentFiles" ensures that projects referencing Microsoft.CodeAnalysis.Common package
       will import everything but content files from Microsoft.CodeAnalysis.Analyzers, specifically, analyzers.
+
+      Disabled in net8.0 until https://github.com/dotnet/roslyn-analyzers/issues/6059 is resolved.
     -->
-    <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="$(MicrosoftCodeAnalysisAnalyzersVersion)" Condition="'$(DotNetBuildFromSource)' != 'true'" PrivateAssets="ContentFiles" />
+    <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="$(MicrosoftCodeAnalysisAnalyzersVersion)" Condition="'$(DotNetBuildFromSource)' != 'true' AND '$(TargetFramework)' != 'net8.0'" PrivateAssets="ContentFiles" />
```
Comment (jjonescz) [inline src/Compilers/Core/Portable/Microsoft.CodeAnalysis.csproj:35]: PublicApi analyzer fails - it suggests changing `IntPtr` to `nint` for net8.0 - https://github.com/dotnet/roslyn-analyzers/issues/6059:

```diff
-static Microsoft.CodeAnalysis.ModuleMetadata.CreateFromImage(System.IntPtr peImage, int size) -> Microsoft.CodeAnalysis.ModuleMetadata!
+static Microsoft.CodeAnalysis.ModuleMetadata.CreateFromMetadata(nint metadata, int size) -> Microsoft.CodeAnalysis.ModuleMetadata!
```

File: eng/generate-compiler-code.ps1 [hunk 9]
```diff
@@ -23,16 +23,17 @@ function Run-LanguageCore($language, $languageSuffix, $languageDir, $syntaxProje
   $errorFilePath = Join-Path $languageDir "Errors\$errorFileName"
   $errorGeneratedFilePath = Join-Path $generatedDir "ErrorFacts.Generated.$($languageSuffix)"
   $targetFramework = "net7.0"
+  $syntaxProjectTargetFramework = if ($language -eq "CSharp") { "net8.0" } else { $targetFramework }
```
Comment (333fred) [inline eng/generate-compiler-code.ps1:26]: We can probably just update all the generators to be 8.0. Nothing keeping them on previous frameworks either.

File: eng/targets/Settings.props [hunk 10]
```diff
@@ -140,7 +149,8 @@
   -->
   <ItemGroup Condition="'$(DotNetBuildFromSource)' != 'true'">
     <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="$(MicrosoftCodeAnalysisNetAnalyzersVersion)" PrivateAssets="all" />
-    <PackageReference Include="Roslyn.Diagnostics.Analyzers" Version="$(RoslynDiagnosticsAnalyzersVersion)" PrivateAssets="all" />
+    <!-- Disabled in net8.0 until https://github.com/dotnet/roslyn-analyzers/issues/6059 is resolved. -->
+    <PackageReference Include="Roslyn.Diagnostics.Analyzers" Version="$(RoslynDiagnosticsAnalyzersVersion)" Condition="'$(TargetFramework)' != 'net8.0'" PrivateAssets="all" />
```
[the remark this answers is not in the document]
Comment (333fred) [reply eng/targets/Settings.props:153]: I think we should just fix the bug.

[the remark this answers is not in the document]
Comment (333fred) [reply eng/targets/Settings.props:153]: > How can we fix the bug before we merge the PR?

I wasn't suggesting this. Merely that, rather than splitting our API files, we should just fix the analyzer (at a later point. Not right now).
